Skin Highlight Controls: highlight Skin Element Groups through controller scripts

absorb-it opened the pull request:
New feature suggestion: allow highlighting of Skin elements with controller commands. This can be especially useful if you activate skin element groups (like one deck) to be controlled with your hardware controller, while other similar element groups can only be controlled from the mixxx GUI.

This is especially useful when you are going to use a 2 deck hardware controller for 4 decks. Instead of one small blinking LED indicating what Deck you are working on, this can be cleanly indicated in the GUI. If you are using a 2-deck controller to play with 4 decks you will have a look on the GUI anyway. 

The example implementation was done for LateNight skins and highlighting of waveform, deckview and deck-equalizers as a group. Maybe it makes sense to make the selection about what to highlight more fine-grained to increase flexibility.

Highlighting can be triggered with `engine.setValue("[Skin]", "highlight_deck[1-4]", true/false)`

I'm open for suggestions.

Commit:
Skin Highlight Controls: initial concept of Skin Control Highlights

allow highlighting of Skin elements with controller command. Useful to
use 2 deck controller for 4 decks.

Commit:
Skin Highlight Controls: preview of new feature

absorb-it commented:
the preview folder contains some screenshots, and will not be part of any 'release' :)

ronso0 commented:
This is indeed a neat helper, thanks!

What I don't like is that we now have borders per se, regardless if there's a script using the highlight controls.
( I see a blank/black border is required to keep un/focused decks / mixer columns / waveforms aligned).
Two simple fixes come to mind, no skin modification required:
1.* use `1` for Off and `2` for On in scripts, `0` means no border (admittedly a bit unintuitive)
2. `[Skin],highlight_deck[1-4]` is `0` by default, and additional helper COs `[Skin],focus_deck[1-4]` set it to `1` (and `2` for the focused deck(s)) as soon as any deck is focused.
similar to the [Mic ducking helper](https://github.com/mixxxdj/mixxx/blob/3c3537038c1c703e78e7d6cbf5b8db893bc7232d/src/skin/skinloader.cpp#L281)

Regarding the alignment of un/focused regions, we may also set a `stacked` layout in the DeckContainer and put a WidgetGroup underneath the deck that takes care of the highlight (via background or border).

What do you think?

absorb-it commented:
Over the day I thought about single-deck controllers or pad-only controllers - it makes sense to implement this in a way that every controller implementation can benefit from it. I think the implementation should be as universal as possible. 

Therefore I suggest the following approach with adding some more controls:

* _highlights_input_allow(true)_  -  globally toggle highlighting based on controller input
* _highlight_deck_X_enable(uint_8)_  -  enable/disable deck- related highlighting with a bitmask:
  * 2⁰ complete outer container
  * 2¹ Spinner
  * 2² Cue/Play
  * 2³ Hotcue
  * 2⁴ Loop
  * 2⁵ Jump
  * 2⁷ Title
* _highlight_mixer_X_enable(uint_8)_  -  enable/disable mixer- related highlighting with a bitmask:
  * 2⁰ complete outer container
  * 2¹ EQ
  * 2² Gain
  * 2³ Volume
* _highlight_preview_X_enable(uint_8)_  -  enable/disable mixer- related highlighting with a bitmask:
  * 2⁰ complete outer container
  * 2¹ Waveform
  * 2² DeckSettingsGrid
* maybe others like _highlight_effectUnit_X_enable(uint_8)_ and _highlight_sampler_X_enable(uint_8)_
* _highlight_group_X_enable(true)_  -  enable/disable highlight group 1-8
* _highlight_group_X_(uint_8) -  configure highlight group 1-8, default 0x15
  * 2⁰ Deck complete outer container (default on)
  * 2¹ Deck Title (default off)
  * 2² Preview Deck Waveform (default on)
  * 2³ Preview Deck SettingsGrid (default off)
  * 2⁴ Mixer Channel complete outer container (default on)
  * 2⁵ Mixer Channel Eq (default off)
  * 2⁶ Mixer Channel Gain (default off)
  * 2⁷ Mixer Channel Volume (default off)
 
It's not clear if all those elements can be easily highlighted, but that's something I can figure out on the way. The approach will allow us
* to enable disable the external highlighting on a global level
* use predefined groups for easy highlighting of combined features related to the controller or
* go into the finest detail with changing the highlighting of small button groups

What do you think?

Commit:
Skin Highlight Controls: added global on/off

added option ([Skin], highlight_input_allow) to globally allow or deny
highlighting from external devices. All changes to layout only applied
if external highlighting is enabled (still requires restart, to be
fixed)

Commit:
Skin Highlight Controls: redraw layout on border change

if global option is required, borders might be activated - redraw is
required

Commit:
Skin Highlight Controls: added tooltip for global option

Commit:
Skin Highlight Controls: added global option preview

including tooltip

absorb-it commented:
Implemented now some 'global' option to allow the external highlighting or not. If this is not activated, the layout is not changed a all. Separate options for all 4 decks are available and formatting is done with qss like
```
#WaveformBox4[highlight_input_allow="1"][highlight_deck="1"] {
    border-color: #82220d;
}
```
Implemented options are now:
```
highlight_input_allow
highlight_deck[1-4]
```
If this seems to be ok like it is I will go trough all styles and try to add some suitable effect for this external highlighting in a final step.

Commit:
Skin Highlight Controls: added Tango Skin implementation

Commit:
Skin Highlight Controls: added Deere Skin implementation

Commit:
Skin Highlight Controls: visual Skin highlight tuning

Commit:
Skin Highlight Controls: remove preview folder

Commit:
Skin Highlight Controls: restored Original and added missing variable

Commit:
Skin Highlight Controls: fixed commment typo

absorb-it commented:
Preview: 
<img width="1596" height="745" alt="LateNight_PaleMoon" src="https://github.com/user-attachments/assets/1d5d0247-c750-4fb0-a49c-0f31c65b898e" />
<img width="1596" height="745" alt="LateNight_Classic" src="https://github.com/user-attachments/assets/b31d5ae7-63b8-4705-810f-68d7c48f3d66" />
<img width="1596" height="745" alt="Deere" src="https://github.com/user-attachments/assets/b0296b69-31fb-47d9-be79-92716cf1a86b" />
<img width="1596" height="745" alt="Tango" src="https://github.com/user-attachments/assets/96b07fc1-bc2e-4fcd-94c9-fe1fd38b7cd0" />

ronso0 commented:
Interesting, this works without defining any `qproperty` in c++ (src/widget/wwidgetgroup.h)?

absorb-it commented:
> Interesting, this works without defining any `qproperty` in c++ (src/widget/wwidgetgroup.h)?

added some properties in 39258cb5f14d12d9e39f9cfd026458c77f877895 , just did it the same way as "highlight" was implemented with 05488db4a3521d22a1ebc0ad89a7dcf871f553e9 - and "just works". Most trouble I had finding some trigger to redraw the widget when Border is shown or not. This got solved with 6e648276abe77b44df13b42afae1e7bac7f3af19

Did I missed something for a clean implementation? Had no issues / warnings on the terminal so far.

ronso0 commented:
> added some properties in https://github.com/mixxxdj/mixxx/commit/39258cb5f14d12d9e39f9cfd026458c77f877895 , just did it the same way as "highlight" was implemented with https://github.com/mixxxdj/mixxx/commit/05488db4a3521d22a1ebc0ad89a7dcf871f553e9 - and "just works".

Whoops, my bad. For some reason I didn't see any c++ changes in the PR diff.
Looks good, will test this and take a closer look soonish :+1:

Swiftb0y commented:
> Tonight I will change the naming of the options from "highlight_deck[1-4]" to "highlight_deck_[1-4]" (additional underscore) to keep this aligned with the other options.

I'd prefer `highlight_GROUP`, where `GROUP` is an already established group. So for instance `highlight_[Channel1]`. This would likely simplify the JS a little.

Commit:
Skin Highlight Controls: renamed deck to channel

renamed highlight_deckX to highlight_[ChannelX] and adapted similar
names

absorb-it commented:
> > Tonight I will change the naming of the options from "highlight_deck[1-4]" to "highlight_deck_[1-4]" (additional underscore) to keep this aligned with the other options.
> 
> I'd prefer `highlight_GROUP`, where `GROUP` is an already established group. So for instance `highlight_[Channel1]`. This would likely simplify the JS a little.

done with 8bc1050fca7f94abcef1f111904b8962399cb71a

Commit:
Skin Highlight Controls: restored file permissons

ronso0 commented on res/skins/Deere/deck.xml:
```diff
@@ -13,8 +13,8 @@
     <Connection>
-      <ConfigKey>[Skin],highlight_deck<Variable name="i"/></ConfigKey>
-      <BindProperty>highlight_deck</BindProperty>
+      <ConfigKey>[Skin],highlight_[Channel<Variable name="i"/>]</ConfigKey>
```
```suggestion
      <ConfigKey>[Skin],highlight_<Variable name="group"/></ConfigKey>
```
; )

Commit:
Skin Highlight Controls: used group variable for settings

Commit:
Skin Highlight Controls: LateNight Palemoon

removed non-required padding between waves in controller-highlight mode

ronso0 commented:
I'll try to do a manual test soonish.

FWIW I'm not a big fan of these 5px borders. Why not stick to 1px like focus borders in all skins?

Commit:
Skin Highlight Controls: changed highlighting style

changed highlighting style to be the same as all other highlightings

Commit:
Skin Highlight Controls: splitted control commands

splitted highlight_channel into separate calls to highlight_deck,
highlight_mixer and highlight_waveform

absorb-it commented:
> FWIW I'm not a big fan of these 5px borders. Why not stick to 1px like focus borders in all skins?

I found the partially bigger borders more obvious and more appealing, but anyway, I'm not sticking to this. And I can change this for my personal skin if I miss it too much :) Changed to standard highlighting in c826d32774caa437e5e89b91ff949303f6ad71c8

absorb-it commented:
> What do you think?

Implemented as suggested in e1509c311ca925fb29557756ad2e3cad103544ef , a lot of options... But as we both agree the a more fine-grained control might be required, this is the way.

absorb-it commented:
Anything than wait that can be done to get this implemented? (bot added stale label)

Best regards, Rene

ronso0 commented:
Sorry this had to wait so long. As ususal priorities shifted and this has been pushed off the table.
Will take a closer look soon (bookmarked, but feel free to ping us again if you have the feeling we forgot about it)

daschuer commented:
This is a now feature that requires a beta period. That's why have moved this to main.
